appliance/postgresql2/state/state.go

```diff
@@ -0,0 +1,1019 @@
+	switch e.Kind {
+	case DiscoverdEventInit:
+		p.handleDiscoverdInit(e)
+	case DiscoverdEventState:
+		p.handleDiscoverdState(e)
+	case DiscoverdEventPeers:
+		p.handleDiscoverdPeers(e)
+	}
+}
+
+func (p *Peer) handleDiscoverdInit(e *DiscoverdEvent) {
+	p.log.Info("discoverd init", "at", "init_discoverd", "peers", len(e.Peers), "state", e.State.State != nil)
+	if p.clusterState != nil {
+		panic("received discoverd init after already initialized")
+	}
+	p.clusterPeers = e.Peers
+	p.decodeState(e)
+	if p.pgOnline != nil {
+		p.evalClusterState()
+	}
+}
+
+func (p *Peer) handleDiscoverdPeers(e *DiscoverdEvent) {
+	p.log.Info("discoverd peers", "at", "discoverd_peers", "peers", len(e.Peers))
+	if p.clusterPeers == nil {
+		panic("received discoverd peers before init")
+	}
+	p.clusterPeers = e.Peers
+	p.evalClusterState()
+}
+
+func (p *Peer) handleDiscoverdState(e *DiscoverdEvent) {
+	log := p.log.New("fn", "handleDiscoverdState", "index", e.State.Index)
+	log.Info("got discoverd state", "state", e.State.State != nil)
+	if p.clusterPeers == nil {
+		panic("received discoverd state before init")
+	}
+	if p.decodeState(e) {
+		p.evalClusterState()
+	} else {
+		log.Info("already have this state", "at", "skip_eval")
+	}
+}
+
+func (p *Peer) decodeState(e *DiscoverdEvent) bool {
+	if e.State.Index > p.clusterStateIndex {
+		p.clusterStateIndex = e.State.Index
+		p.clusterState = e.State.State
+		return true
+	}
+	return false
+}
+
+// evalLater triggers a cluster state evaluation after delay has elapsed
+func (p *Peer) evalLater(delay time.Duration) {
+	if p.retryCh != nil {
+		p.retryCh <- struct{}{}
+		return
+	}
+	time.AfterFunc(delay, p.triggerEval)
+}
+
+func (p *Peer) applyConfigLater(delay time.Duration) {
+	if p.retryCh != nil {
+		p.retryCh <- struct{}{}
+		return
+	}
+	time.AfterFunc(delay, p.triggerApplyConfig)
+}
+
+func (p *Peer) triggerEval() {
+	select {
+	case p.evalStateCh <- struct{}{}:
+	default:
+		// if we can't send to the channel, there is already a pending state
+		// evaluation, as it is buffered
+	}
+}
+
+func (p *Peer) triggerApplyConfig() {
+	select {
+	case p.applyConfCh <- struct{}{}:
+	default:
+		// if we can't send to the channel, there is already a pending
+		// configuration, as it is buffered
+	}
+}
+
+func (p *Peer) moving() {
+	if !p.atRest() {
+		return
+	}
+	p.log.Debug("moving")
+
+	// Set up a channel that the select in the run loop can use to detect
+	// when there is no more work to do.
+	p.workDoneCh = make(chan struct{}, 1)
+	p.workDoneCh <- struct{}{}
+}
+
+func (p *Peer) rest() {
+	p.log.Debug("at rest", "fn", "rest")
+	p.workDoneCh = nil
+	if p.restCh != nil {
+		p.restCh <- struct{}{}
+	}
+}
+
+func (p *Peer) atRest() bool {
+	return p.workDoneCh == nil
+}
+
+// Examine the current cluster state and determine if new actions need to be
+// taken. For example, if we're the primary, and there's no sync present, then
+// we need to declare a new generation.
+func (p *Peer) evalClusterState() {
+	p.moving()
```

Comment:
This is used by the simulator/test suite to figure out when all of the triggered work has been completed. At the start of an evaluation, moving will get called, and one or more evaluation iterations may happen. When there is no more work to do, the select will end up receiving from `workDoneCh` and trigger the resting state.